x/concentrated-liquidity/math/tick.go

```diff
@@ -206,7 +210,64 @@ func CalculatePriceToTick(price sdk.Dec) (tickIndex int64) {
 	// * taking the bucket index of the smallest price in this tick
 	// * adding to it the number of ticks "completely" filled by the current spacing
 	// the latter is the truncation of the division above
-	// TODO: This should be rounding down?
-	tickIndex = geoSpacing.initialTick + ticksFilledByCurrentSpacing.SDKDec().RoundInt64()
-	return tickIndex
+	tickIndex = geoSpacing.initialTick + ticksFilledByCurrentSpacing.SDKDec().TruncateInt64()
+
+	// We get the price corresponding to our calculated bucket index to compare against the input price.
+	curPrice, err := TickToPrice(tickIndex)
+	if err != nil {
+		return 0, err
+	}
+
+	// We first handle cases related to min and max tick independently here, as in either case checking both the tick above and below is not possible.
+	if curPrice.Equal(types.MinSpotPrice) && curPrice.LTE(price) {
+		// We assume that the error to the downside is caught by input price bound checks, as it
+		// would be below the minimum allowed spot price.
+		return types.MinTick, nil
+	} else if curPrice.Equal(types.MaxSpotPrice) && curPrice.GTE(price) {
+		// We assume that the error to the upside is caught by input price bound checks, so we only
+		// check the price in the tick below.
+		tickBelowPrice, err := TickToPrice(tickIndex - 1)
+		if err != nil {
+			return 0, err
+		}
+
+		// If price is lower than the price in the tick below, our error assumption is violated so we panic.
+		if price.LT(tickBelowPrice) {
+			panic(fmt.Sprintf("price %s is outside of error bounds for tick %d", price, tickIndex))
+		}
+
+		// If price falls in the bucket below `tickIndex`, return that bucket index.
+		// Otherwise, return the current tick index.
+		if price.GTE(tickBelowPrice) && price.LT(curPrice) {
+			return tickIndex - 1, nil
+		}
```

Comment:
I think we should make a new SqrtPriceToTick function, that does the squaring internally, then does this comparison on the derived sqrt prices. As the problem is our price definition doesn't match due to errors.

(I don't think this on its own will save us from anything aside from the erroneous bucket definition due to not truncating)